### PR TITLE
Surface mesh topology shortest cycle

### PR DESCRIPTION
**Pull request by gdamiand**
_Please use the following template to help us managing pull requests._

## Summary of Changes

Add a method that computes the smallest non contractible cycle in the surface mesh topology package.

## Release Management

* Affected package(s): Surface mesh topology
* Small Feature : [Shortest_noncontractible_cycle](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Shortest_noncontractible_cycle) -- Pre-approved Gdamiand 17:01, 14 April 2020 (CEST) 

**TODO:**
- [x] @sloriot Check if branch must be filtered (size + duplicated commits)

### CODE REVIEW

**Comment by gdamiand** [comment]
@maxGimeno This PR is ready for testing.

File: Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt [hunk 1]
```diff
@@ -8,16 +8,35 @@ namespace CGAL {
...
+
 
 \section SMTopology Introduction
 
-Given a curve drawn on a surface one can ask if the curve can be continuously deformed to a point (i.e. a zero length curve). In other words, does there exist a continuous sequence of curves on the surface that starts with the input curve and ends to a point? Curves that deform to a point are said <em>contractible</em>. Any curve on a sphere is contractible but this is not true for all curves on a torus or on a surface with more complicated topology. The algorithms in this package are purely topological and do not assume any geometry on the input surface. In particular, the surface is not necessarily embedded in a Euclidean space.
+All the computations in this package, either for the shortest non-contractible curve or for the homotopy tests,  are performed on a input surface represented as a model of `CombinatorialMap` or any model of `FaceGraph`. Note that combinatorial maps are based on darts and FaceGraphs are based on halfedges. To avoid repetitions we use the terms darts and halfedges interchangeably in the sequel.
+The input surface is supposed to be connected and orientable.
```
Comment (sloriot) [inline Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt:17]: you mean that the number of connected component = 1?

File: Surface_mesh_topology/examples/Surface_mesh_topology/CMakeLists.txt [hunk 2]
```diff
@@ -25,24 +25,54 @@ endif()
...
+  path_homotopy_with_sm_and_polyhedron.cpp
   path_homotopy_with_symbols_2.cpp
-  surface_mesh_topology_with_sm_and_polyhedron.cpp
+  path_homotopy_with_symbols.cpp
+  shortest_noncontractible_cycle_2.cpp
+  shortest_noncontractible_cycle.cpp
+  unsew_edgewidth_repeatedly.cpp
+  )
+
+set(HEADER_FILES
```
Comment (sloriot) [inline Surface_mesh_topology/examples/Surface_mesh_topology/CMakeLists.txt:44]: really?

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Facewidth.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  using Dart_const_handle=typename Local_map::Dart_const_handle;
+  using size_type        = typename Local_map::size_type;
+
+  using Path             = CGAL::Surface_mesh_topology::Path_on_surface<Mesh>;
+  using SNC              = Shortest_noncontractible_cycle<Local_map, false>;
+
+  Facewidth(const Mesh& amesh, bool display_time=false):
+    m_snc_to_find_facewidth(nullptr)
+  {
+    CGAL::Timer t;
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Facewidth.h:47]: Is it expected to always have the timer?

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+         operator() (const HEG& heg, Dart_const_handle dh)
+ * Index_from_halfedge_descriptor<Mesh>::run(m, h)
+ * Halfedge_descriptor_from_index<Mesh>::run(m, i)
+ * Is_index_used<Mesh>::run(m, i)
+*/
+////////////////////////////////////////////////////////////////////////////////
+namespace CGAL {
+////////////////////////////////////////////////////////////////////////////////
+template<typename P>
+class Surface_mesh;
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Functors_for_face_graph_wrapper.h:32]: why is this needed?

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Generic_map_selector.h [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+#include <CGAL/Generalized_map.h>
+#include <CGAL/Linear_cell_complex_for_combinatorial_map.h>
+#include <CGAL/Linear_cell_complex_for_generalized_map.h>
+#include <CGAL/Polygonal_schema.h>
+
+namespace CGAL {
+namespace Surface_mesh_topology {
+namespace internal {
+
+  struct Items_for_shortest_noncontractible_cycle
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Generic_map_selector.h:29]: namespace shall not  imply indentation

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+    int vertex_index=0;
+    size_type vertex_visited;
+    try
+    {
+      vertex_visited=get_local_map().get_new_mark();
+    }
+    catch (typename Local_map::Exception_no_more_available_mark)
+    {
+      std::cerr<<"No more free mark, exit."<<std::endl;
+      exit(-1);
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:276]: This means killing the user application, using an exception or just not catching it seems more user friendly to me.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+      { get_local_map().template set_attribute<0>
+          (it, get_local_map().template create_attribute<0>(-1)); }
+    }
+  }
+
+  void initialize_vertex_info()
+  {
+    for (auto it=get_local_map().template attributes<0>().begin(),
+           itend = get_local_map().template attributes<0>().end(); it != itend; ++it)
+    { get_local_map().template info_of_attribute<0>(it)=-1; }
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:230]: @gdamiand  are you also using this strange indentation style are is it only due to the new author?

Comment (sloriot) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:230]: If needed I can run a astyle on the files to fix those. (it will also fix the extra indentation due to namespace)

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+      get_local_map().unmark(it, edge_deleted);
+      get_local_map().unmark(it, face_deleted);
+    }
+
+    get_local_map().free_mark(tested);
+    get_local_map().free_mark(edge_deleted);
+    get_local_map().free_mark(face_deleted);
+  }
+
+  Dart_handle nonhole_dart_of_same_edge(Dart_handle dh)
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:497]: non_hole?

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+                     typename WeightFunctor::Weight_t* length,
+                     const typename WeightFunctor::Weight_t* max_length,
+                     const WeightFunctor& wf)
+  {
+    std::vector<typename WeightFunctor::Weight_t> distance_from_root;
+    m_spanning_tree.clear();
+    m_trace_index.clear();
+    initialize_vertex_info();
+    compute_spanning_tree(root, m_spanning_tree, distance_from_root, m_trace_index, wf);
+    compute_noncon_edges(m_spanning_tree, m_noncon_edges);
```
Comment (sloriot) [inline Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:543]: I guess `compute_noncon_edges` is not public?

File: Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt [hunk 10]
```diff
@@ -194,7 +264,7 @@ It can be proven that the canonical form is uniquely defined and only depends on
 
 \section Implementation History
 
-The code was developed in 2018 by Guillaume Damiand and Francis Lazarus. Felix Castillon contributed to the extension of the homotopy test to the case of surfaces with boundaries.
+The code was developed in 2018 by Guillaume Damiand and Francis Lazarus. Felix Castillon contributed to the extension of the homotopy test to the case of surfaces with boundaries. Thien Hoang added methods to compute shortest  non-contractible cycles and face-width as part of the program Google Summer of Code 2019.
```
Comment (sloriot) [inline Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt:267]: no mention to edge-width?

**Comment by sloriot** [comment]
```
Surface_mesh_topology/examples/Surface_mesh_topology/draw_facewidth.h:23:32: warning: unused parameter ‘alcc’ [-Wunused-parameter]
Surface_mesh_topology/examples/Surface_mesh_topology/draw_facewidth.h:23:70: warning: unused parameter ‘dh’ [-Wunused-parameter]
Surface_mesh_topology/examples/Surface_mesh_topology/draw_facewidth.h:56:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Dart<2, CGAL::CMap_linear_cell_complex_storage_1<2, 3, CGAL::Linear_cell_complex_traits<3, CGAL::Epick>, CGAL::Linear_cell_complex_min_items, std::allocator<int> >, CGAL::Void, CGAL::Boolean_tag<false> >, std::allocator<CGAL::Dart<2, CGAL::CMap_linear_cell_complex_storage_1<2, 3, CGAL::Linear_cell_complex_traits<3, CGAL::Epick>, CGAL::Linear_cell_complex_min_items, std::allocator<int> >, CGAL::Void, CGAL::Boolean_tag<false> > >, CGAL::Default, CGAL::Default>, true>, std::allocator<CGAL::internal::CC_iterator<CGAL::Compact_container<CGAL::Dart<2, CGAL::CMap_linear_cell_complex_storage_1<2, 3, CGAL::Linear_cell_complex_traits<3, CGAL::Epick>, CGAL::Linear_cell_complex_min_items, std::allocator<int> >, CGAL::Void, CGAL::Boolean_tag<false> >, std::allocator<CGAL::Dart<2, CGAL::CMap_linear_cell_complex_storage_1<2, 3, CGAL::Linear_cell_complex_traits<3, CGAL::Epick>, CGAL::Linear_cell_complex_min_items, std::allocator<int> >, CGAL::Void, CGAL::Boolean_tag<false> > >, CGAL::Default, CGAL::Default>, true> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/examples/Surface_mesh_topology/edgewidth_lcc.cpp:15:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/examples/Surface_mesh_topology/edgewidth_surface_mesh.cpp:15:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/examples/Surface_mesh_topology/shortest_noncontractible_cycle_2.cpp:113:20: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/examples/Surface_mesh_topology/shortest_noncontractible_cycle.cpp:16:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/examples/Surface_mesh_topology/unsew_edgewidth_repeatedly.cpp:120:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/include/CGAL/Face_graph_wrapper.h:589:25: warning: no return statement in function returning non-void [-Wreturn-type]
Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Iterators_for_face_graph_wrapper.h:61:5: warning: control reaches end of non-void function [-Wreturn-type]
Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:298:33: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<double>::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:298:33: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<double, std::allocator<double> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:298:33: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<unsigned int>::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:331:46: warning: unused parameter ‘wf’ [-Wunused-parameter]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:121:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:143:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:192:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:254:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:265:10: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:272:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:283:10: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:290:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:300:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<CGAL::Point_3<CGAL::Epick>, std::allocator<CGAL::Point_3<CGAL::Epick> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:305:61: warning: statement has no effect [-Wunused-value]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:306:61: warning: statement has no effect [-Wunused-value]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:316:56: warning: statement has no effect [-Wunused-value]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:317:61: warning: statement has no effect [-Wunused-value]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:328:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<CGAL::Point_3<CGAL::Epick>, std::allocator<CGAL::Point_3<CGAL::Epick> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:330:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<CGAL::Point_3<CGAL::Epick>, std::allocator<CGAL::Point_3<CGAL::Epick> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:332:22: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<CGAL::Point_3<CGAL::Epick>, std::allocator<CGAL::Point_3<CGAL::Epick> > >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:365:23: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:380:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<unsigned int>::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/shortest_noncontractible_cycle_tests.cpp:99:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp:169:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp:81:33: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
```

**Comment by sloriot** [comment]
@maxGimeno even if the package has some warnings and @gdamiand does not have the time to fix them, this PR can still go into integration as it will impact only one package.

**Comment by gdamiand** [comment]
> @maxGimeno even if the package has some warnings and @gdamiand does not have the time to fix them, this PR can still go into integration as it will impact only one package.

Thanks @sloriot ; will work on the warning this afternoon.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h (hunk 9 above)
Comment (gdamiand) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:543]: yes.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h (hunk 8 above)
Comment (gdamiand) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:497]: Undocumented internal method; thus I suppose not really important.

File: Surface_mesh_topology/doc/Surface_mesh_topology/CGAL/Path_on_surface.h [hunk 11]
```diff
@@ -39,12 +39,12 @@ namespace Surface_mesh_topology {
     /// clears this path.
     void clear();
 
-    /// returns `true` iff `hd` can be added at the end of this path.
-    bool can_be_pushed(halfedge_descriptor hd) const;
+    /// returns `true` iff `hd` can be added at the end of this path. If `flip` is true, `hd`'s direction is reversed before checking
+    bool can_be_pushed(halfedge_descriptor hd, bool flip=false) const;
 
-    /// adds `hd` at the end of this path.
+    /// adds `hd` at the end of this path. If `flip` is true, the reverse of `hd` is considered.
```
[the remark this answers is not in the document]
Comment (gdamiand) [reply Surface_mesh_topology/doc/Surface_mesh_topology/CGAL/Path_on_surface.h:45]: Done.

File: Surface_mesh_topology/doc/Surface_mesh_topology/Concepts/WeightFunctor.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  \cgalHasModel \link CGAL::Surface_mesh_topology::Unit_weight_functor `CGAL::Surface_mesh_topology::Unit_weight_functor`\endlink
+  \cgalHasModel \link CGAL::Surface_mesh_topology::Euclidean_length_weight_functor `CGAL::Surface_mesh_topology::Euclidean_length_weight_functor<Mesh>`\endlink
+*/
+class WeightFunctor {
+public:
+/// \name Public types
+/// @{
+
+  /*!
+    %halfedge_descriptor type. A handle to Dart for combinatorial/generalized maps, or a halfedge descriptor for models of FaceGraph.
```
[the remark this answers is not in the document]
Comment (gdamiand) [reply Surface_mesh_topology/doc/Surface_mesh_topology/Concepts/WeightFunctor.h:17]: Done.

File: Installation/CHANGES.md [hunk 13]
```diff
@@ -1,6 +1,14 @@
 Release History
 ===============
 
+[Release 5.2] (https://github.com/CGAL/cgal/releases/tag/releases%2FCGAL-5.2)
+-----------
+
+### Surface Mesh Topology
+
+- Added new functionalities to compute shortest non contractible cycles on a surface, with or without weights on edges.
```
[the remark this answers is not in the document]
Comment (gdamiand) [reply Installation/CHANGES.md:9]: Done.

File: Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt (hunk 10 above)
Comment (gdamiand) [reply Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt:267]: Done.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h (hunk 6 above)
Comment (gdamiand) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:276]: Done.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h (hunk 8 above)
Comment (gdamiand) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:496]: undocumented internal method; not really important.

File: Surface_mesh_topology/doc/Surface_mesh_topology/PackageDescription.txt [hunk 14]
```diff
@@ -21,7 +21,7 @@
...
-\cgalPkgDesc{This package provides methods for testing if two (closed) paths on a combinatorial surface are homotopic. The user can choose between free homotopy and homotopy with fixed endpoints. A contractibility test is also provided; it amounts to test homotopy with the constant path. After an initialization step that takes linear time in the size of the input surface, the homotopy tests are performed in linear time in the size of the input curves.}
+\cgalPkgDesc{This package provides methods for testing if two (closed) paths on a combinatorial surface are homotopic and for computing edge-width and face-width of the mesh. The user can choose between free homotopy and homotopy with fixed endpoints. A contractibility test is also provided; it amounts to test homotopy with the constant path. After an initialization step that takes linear time in the size of the input surface, the homotopy tests are performed in linear time in the size of the input curves. While computing edge-width or face-width, the user can opt to take geometric distances between vertices into consideration, or disregard them so as to make the mesh unweighted.}
```
[the remark this answers is not in the document]
Comment (gdamiand) [reply Surface_mesh_topology/doc/Surface_mesh_topology/PackageDescription.txt:24]: Done.

File: Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h (hunk 7 above)
Comment (gdamiand) [reply Surface_mesh_topology/include/CGAL/Surface_mesh_topology/internal/Shortest_noncontractible_cycle.h:230]: Done.

File: Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt [hunk 15]
```diff
@@ -8,16 +8,35 @@ namespace CGAL {
 \cgalAutoToc
 \author Guillaume Damiand and Francis Lazarus
 
-This package provides an algorithm to test if a curve on a combinatorial surface can be continuously deformed into another curve.
+This package provides a toolbox for manipulating curves on a combinatorial surface from the topological viewpoint. Two main functionalities are proposed. One is concerned with the computation of shortest curves that cannot be continuously deformed to a point. This includes the computation of the so-called edge-width and face-width of the vertex-edge graph of a combinatorial surface. The other functionality is concerned with the homotopy test for deciding if two given curves on a combinatorial surface can be continuously deformed one into the other.
```
[the remark this answers is not in the document]
Comment (gdamiand) [reply Surface_mesh_topology/doc/Surface_mesh_topology/Surface_mesh_topology.txt]: Done.

**Comment by sloriot** [comment]
```
Surface_mesh_topology/examples/Surface_mesh_topology/shortest_noncontractible_cycle.cpp:16:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp:169:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp:81:33: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::size_t’ {aka ‘long unsigned int’} [-Wsign-compare]
```

**Comment by gdamiand** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-134.shtml

**Comment by sloriot** [comment]
@gdamiand there are some `.off` that are added by this branch that are not used. Can I remove them?

**Comment by gdamiand** [comment]
> @gdamiand there are some `.off` that are added by this branch that are not used. Can I remove them?

If you want. They can be used by examples that take an off as input; but this is not mandatory...

File: Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+    res=false;
+  }
+
+  Marked_weight_functor<LCC_CM> wf(map, mark);
+  CGAL::Surface_mesh_topology::Curves_on_surface_topology<LCC_CM> cst(map);
+
+  auto cycle2=cst.compute_shortest_non_contractible_cycle(wf);
+  if (cycle2.length()!=nbedges)
+  {
+    std::cout<<"[ERROR] in test_weighted for double-torus-2-d.off: the length"
```
Comment (sloriot) [inline Surface_mesh_topology/test/Surface_mesh_topology/test_shortest_cycle_non_contractible.cpp:122]: hardcoded name seems to be a left over (same remark a few lines below)

**Comment by sloriot** [comment]
namely:
```
Surface_mesh_topology/examples/Surface_mesh_topology/data/double-torus-2-a.off
Surface_mesh_topology/test/Surface_mesh_topology/data/double-torus-2-c.off
Surface_mesh_topology/test/Surface_mesh_topology/data/double-torus-2-d.off
Surface_mesh_topology/test/Surface_mesh_topology/data/elephant-with-holes.off
Surface_mesh_topology/test/Surface_mesh_topology/data/obj1.off
Surface_mesh_topology/test/Surface_mesh_topology/data/obj1-with-holes.off
```

**Comment by sloriot** [comment]
For some reason there were some duplicated commits from the initial feature branch. I cleaned that + removed the file above. The diff with your branch (I pushed a backup copy on your remote `gsoc2019-nccycles-thien_before_28042020_filtering`) is limited to the files I removed.

**Comment by sloriot** [comment]
I can merge the branch now if you have no more things to do in it. I'm still planning to look at the user manual before the release but I can open a dedicated issue for that.

**Comment by gdamiand** [comment]
Ok @sloriot ; nothing more to do from my side. Thanks !

**Comment by sloriot** [comment]
@gdamiand There hasn't been a news entry for the Surface mesh topology package. You should consider write one. See the lastest one [here](https://www.cgal.org/2020/04/20/Optimal_bounding_box/).

**Comment by gdamiand** [comment]
@sloriot Something like that https://github.com/CGAL/cgal-web/pull/6 ?
